File: language/src/main/java/brj/nodes/FnRootNode.java
package brj.nodes;

import brj.BridjeLanguage;
import com.oracle.truffle.api.frame.FrameDescriptor;
import com.oracle.truffle.api.frame.VirtualFrame;
import com.oracle.truffle.api.instrumentation.GenerateWrapper;
import com.oracle.truffle.api.instrumentation.InstrumentableNode;
import com.oracle.truffle.api.instrumentation.ProbeNode;
import com.oracle.truffle.api.nodes.ExplodeLoop;
import com.oracle.truffle.api.nodes.RootNode;

@GenerateWrapper
public class FnRootNode extends RootNode implements InstrumentableNode {

    @Children
    private final WriteLocalNode[] writeArgNodes;

    @Child
    private ExprNode exprNode;

    protected FnRootNode(FnRootNode copy) {
        this(copy.getLanguage(BridjeLanguage.class), copy.getFrameDescriptor(),
            copy.writeArgNodes, copy.exprNode);
    }

    public FnRootNode(BridjeLanguage lang, FrameDescriptor frameDescriptor,
                      WriteLocalNode[] writeArgNodes, ExprNode exprNode) {
        super(lang, frameDescriptor);
        this.writeArgNodes = writeArgNodes;
        this.exprNode = exprNode;
    }

    @Override
    @ExplodeLoop
    public Object execute(VirtualFrame frame) {
        for (WriteLocalNode writeArgNode : writeArgNodes) {
            writeArgNode.execute(frame);
        }

        return exprNode.execute(frame);
    }


    @Override
    public boolean isInstrumentable() {
        return true;
    }

    @Override
    public WrapperNode createWrapper(ProbeNode probe) {
        return new FnRootNodeWrapper(this, this, probe);
    }
}

File: language/src/main/kotlin/brj/nodes/FnRootNode.kt
package brj.nodes

import brj.BridjeNode
import brj.BridjeLanguage
import brj.runtime.BridjeRecord
import com.oracle.truffle.api.frame.FrameDescriptor
import com.oracle.truffle.api.frame.VirtualFrame
import com.oracle.truffle.api.nodes.RootNode

class FnRootNode(
    language: BridjeLanguage,
    frameDescriptor: FrameDescriptor,
    private val paramCount: Int,
    private val hasCapturedValues: Boolean = false,
    @field:Child private var bodyNode: BridjeNode
) : RootNode(language, frameDescriptor) {

    override fun execute(frame: VirtualFrame): Any? {
        val args = frame.arguments
        val paramStartIndex = if (hasCapturedValues) 1 else 0
        val argCount = args.size - paramStartIndex
        repeat(paramCount) { i ->
            frame.setObject(i, if (i < argCount) args[paramStartIndex + i] else BridjeRecord.EMPTY)
        }
        return bodyNode.execute(frame)
    }
}
